File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.java
class Solution {
    public String reverseWords(String s) {
        s = s.trim();
        int i = s.length() - 1 , j = s.length();
        StringBuilder ans  = new StringBuilder();

        while( i >= 0){
            if(s.charAt(i) == ' '){
                String substr = s.substring(i+1 , j).trim();
                if(substr.length() > 0){
                         ans.append(substr);
                        ans.append(" ");
                }
               j =i;
            }


            i--;
        }
        ans.append(s.substring(i + 1 , j));
        return ans.toString();
    }
}

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.cpp
class Solution {
public:
    string reverseWords(string s) {
        int ptr = s.size() - 1;
        string word = "";

        // skip trailing spaces
        while (ptr >= 0 && s[ptr] == ' ') ptr--;

        int end = ptr;

        while (ptr >= 0) {
            if (s[ptr] == ' ') {
                if (end > ptr) {
                    if (!word.empty()) word.push_back(' ');
                    word += s.substr(ptr + 1, end - ptr);
                }
                while (ptr >= 0 && s[ptr] == ' ') ptr--;
                end = ptr;
            } else {
                ptr--;
            }
        }

        // first word
        if (end >= 0) {
            if (!word.empty()) word.push_back(' ');
            word += s.substr(0, end + 1);
        }

        return word;
    }
};
